dolphin/drivers/huawei/oceanstor/oceanstor.py

```diff
@@ -0,0 +1,181 @@
+        self.sector_size = int(storage['SECTORSIZE'])
+
+        total_cap = int(storage['TOTALCAPACITY']) * self.sector_size
+        used_cap = int(storage['USEDCAPACITY']) * self.sector_size
+        free_cap = int(storage['userFreeCapacity']) * self.sector_size
+
+        s = {
+            'name': 'OceanStor',
+            'vendor': 'Huawei',
+            'description': 'Huawei OceanStor Storage',
+            'model': storage['NAME'],
+            'status': status,
+            'serial_number': storage['ID'],
+            'firmware_version': '',
+            'location': storage['LOCATION'],
+            'total_capacity': total_cap,
+            'used_capacity': used_cap,
+            'free_capacity': free_cap
+        }
+        LOG.info("get_storage(), successfully retrieved storage details")
+        return s
+
+    def list_pools(self, context):
+        try:
+            # Get list of OceanStor pool details
+            pools = self.client.get_all_pools()
+
+            pool_list = []
+            for pool in pools:
+                # Get pool status
+                status = constants.PoolStatus.OFFLINE
+                if pool['HEALTHSTATUS'] == consts.STATUS_HEALTH \
+                        and pool['RUNNINGSTATUS'] == \
+                        consts.STATUS_POOL_ONLINE:
+                    status = constants.PoolStatus.NORMAL
+
+                # Get pool storage_type
+                storage_type = None
+                if pool.get('USAGETYPE') == consts.BLOCK_STORAGE_POOL_TYPE:
+                    storage_type = constants.StorageType.BLOCK
+                if pool.get('USAGETYPE') == consts.FILE_SYSTEM_POOL_TYPE:
+                    storage_type = constants.StorageType.FILE
+
+                total_cap = \
+                    int(pool['USERTOTALCAPACITY']) * self.sector_size
+                used_cap = \
+                    int(pool['USERCONSUMEDCAPACITY']) * self.sector_size
+                free_cap = \
+                    int(pool['USERFREECAPACITY']) * self.sector_size
+
+                p = {
+                    'name': pool['NAME'],
+                    'storage_id': self.storage_id,
+                    'original_id': pool['ID'],
+                    'description': 'Huawei OceanStor Pool',
+                    'status': status,
+                    'storage_type': storage_type,
+                    'total_capacity': total_cap,
+                    'used_capacity': used_cap,
+                    'free_capacity': free_cap,
+                }
+                pool_list.append(p)
+
+            return pool_list
+
+        except Exception as err:
+            LOG.error(
+                "Failed to get pool metrics from OceanStor: {}".format(err))
+            raise exception.StorageBackendException(
+                reason='Failed to get pool metrics from OceanStor')
+
+    def list_volumes(self, context):
+        try:
+            # Get all volumes in OceanStor
+            volumes = self.client.get_all_volumes()
```

Comment:
Hi @joseph-v please use pagination here, you cannot get all volumes at once if too much volumes exist.

Reply:
Done